nixos/trafficserver: avoid builtins.readFile

rnhmjoj opened the pull request:
###### Motivation for this change

Using builtins.readFile to load upstream defaults is a clever trick, but
it's not allowed in restricted evaluation mode: which means it fails on
Hydra, for example. Besides - in Nixpkgs - depending on derivation as
inputs is considered bad practice and should be avoided.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested via `nixosTests.trafficserver` (the inputs seems to be identical)
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc: @lukegb @vcunat

midchildan reviewed (approved):
Thanks.

> depending on derivation as inputs is considered bad practice and should be avoided.

Embarrassingly, it's the first I've heard of this issue. I assume the reason for this is to avoid triggering builds when running checks and generating docs for the module?

rnhmjoj commented:
The main reason is that Hydra performs all evaluations in restricted mode (for security and purity reasons), so the module fails to evaluate. This means the `nixosTests.trafficserver` was never run automatically and (hypotetically) it could not have been included into the `tested` jobset.

rnhmjoj commented:
By the way, what I'm proposing here is the most basic solution, but not very clean.
So, if you have other ideas, maybe without including the content of the yaml encoded as Nix, do tell.
For example, does leaving the settings empty implies the upstream defauls are used or result in an error?

midchildan commented:
Maybe converting the YAML configuration to JSON beforehand and importing it as a file bundled in Nixpkgs would make future updates easier and also avoid creating a derivation? But it's currently only a few lines of configuration, so it could be dealt with at a later point in time.

Leaving the settings unspecified shouldn't result in an error, but would still be undesirable because it would deviate from the defaults. For `logging.yaml`, it would disable logging. As for `ip_allow.yaml`, it would allow anyone to send requests that manipulate HTTP caches on the server.

rnhmjoj committed:
nixos/trafficserver: avoid input from derivation

Using builtins.readFile to load upstream defaults is a clever trick, but
it's not allowed in restricted evaluation mode: which means it fails on
Hydra, for example. Besides - in Nixpkgs - depending on derivation as
inputs is considered bad practice and should be avoided.

rnhmjoj commented:
Ok, I converted the YAML to JSON and put the two files next to the `trafficserver.nix` (now `default.nix`).
This way it should be easier to regenerate the file when trafficserver is update.

rnhmjoj commented:
@GrahamcOfBorg test trafficserver

midchildan commented:
Looks perfect. The old `trafficserver.nix` file is still remaining though, so I think it's good to go after that's deleted.

rnhmjoj commented:
Oops, fixed. I was in fact wondering why git didn't realise the file has been renamed.
